File: Week10/P1969.py

def greedy(N, M, dna_list):
        
    #결과 저장리스트
    result_list = []

    for i in range(M):
        # 빈도수 저장 딕셔너리 -> 각 문자의 빈도수   
        frequency = {}
        for j in range(N):
            if dna_list[j][i] in frequency:
                frequency[dna_list[j][i]] += 1
            else:
                frequency[dna_list[j][i]] = 1

        #가장 빈도수 높은 문자 찾기             
        most_common_char = None
        max_count = 0
        for char, count in frequency.items():
            # 사전순으로 배치 해야 함...!!
            if count > max_count or (count == max_count and (most_common_char is None or char < most_common_char)):
                most_common_char = char
                max_count = count

        result_list.append(most_common_char)

    hamming_distance = 0
    for i in range(M):
        for j in range(N):
            if dna_list[j][i] != result_list[i]:
                 hamming_distance += 1

        

    result_str = "".join(result_list)  # 리스트를 문자열로 변환

    return result_str, hamming_distance


#dna 수, 문자열의 길이 
N, M  = map(int, input().split())

#dna 배열 초기화 
dna_list = []
for i in range(N):
    # 공백이 아닌 문자열 하나하나 구분할 때에는 strip() 사용!!!
    row = list(input().strip())
    dna_list.append(row)
    

result, hamming_distance = greedy(N, M, dna_list)
print(result)
print(hamming_distance)

File: Week10/P1969.java

import java.util.Scanner;


public class P1969{

    //변수 저장 클래스 
    static class Result{
        String dna;
        int hamming_distance;

        public Result(String dna, int hamming_distance){
            this.dna = dna;
            this.hamming_distance = hamming_distance;
        }

    }
    public static Result get_hamming_distance(int N, int M, char dna_list[][]){

        // 결과 저장 배열
        StringBuilder result = new StringBuilder();
        int hamming_distance = 0;

        for(int i = 0; i <M; i++){

            // 빈도수 저장 베열 초기화 
            int [] frequency = new int[26]; //A~Z에 해당하는 빈도수 

            for (int j = 0; j <N; j++){
                char c = dna_list[j][i];
                frequency[c-'A']++; //문자에 해당하는 숫자에 빈도수 저장 (문자를 숫자로 변환)
            }
            char most_common_char = 'A';
            int max_count = 0;

            //가장 빈도수 높은 문자를 저장 후, 사전순! 
            for (int k = 0; k < 26; k++){
                // k+'A' -> 숫자를 문자로 변환! 
                if( frequency[k] > max_count || ((frequency[k] == max_count) && k +'A' < most_common_char)){
                    most_common_char = (char)(k+'A');
                    max_count = frequency[k];
                }
            }
            //가장 빈도수 높고, 사전순으로 앞에 있는 문자 결과 문자열에 추가!
            result.append(most_common_char);
            
            for (int j = 0; j < N; j++) {
                if(dna_list[j][i] != most_common_char){
                    hamming_distance += 1;
                }
            }

        }
        return new Result(result.toString(), hamming_distance);
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        int N = sc.nextInt();
        int M = sc.nextInt(); 
        sc.nextLine(); // 버퍼 비우기

        // dna 리스트 
        char [][] dna_list = new char[N][M];

        for (int i = 0; i < N; i++){
            String row = sc.nextLine();
            dna_list[i] = row.toCharArray(); //문자열을 문자배열로 변경해서 저장

        }
        Result result = get_hamming_distance(N, M, dna_list);

        System.out.println(result.dna);
        System.out.println(result.hamming_distance);
    }
}